Introduce initial support for Hedera

mehcode opened the pull request:
## Context (issues, jira)

## Description / Usage

- Support for scanning and synching hedera accounts
- No support yet for sending or receiving hbar

## Expectations

- [ ] **Test coverage: The changes of this PR are covered by test.** Unit test were added with mocks when depending on a backend/device.
- [ ] **No impact: The changes of this PR have ZERO impact on the userland.** Meaning, we can use these changes without modifying LLD/LLM at all. It will be a "noop" and the maintainers will be able to bump it without changing anything.

mehcode committed:
Add support for Hedera

 - designate a custom deriviation as the bolos app does not support hardened paths

 - define methods for querying accounts by key and balance with the mirror and node respectively

 - implement scanAccounts and sync using the mirror and node APIs

 + operations should always hold a positive value (even for "send" operations)

 + fake blockHeight to show operations as confirmed

 + add the concept of a virtualAddress and use for Hedera,
   a set virtual address will directly render it during receive and
   warn the user that it cannot be verified

gre commented on src/data/icons/svg/HBAR.svg:
```diff
@@ -0,0 +1,25 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
+<svg
+   id="Layer_1"
```
we need cleaner .svg that don't have "id" and "defs" in it, because it wouldn't work when converted to components (like React Native components)

FabriceDautriat replied:
@gre we decided to have it designed internally
(cc @rbeyon-ledger )

gre commented on src/families/hedera/api/mirror.ts:
```diff
@@ -0,0 +1,108 @@
+  network({
+    type: "get",
+    url: URL.format({
+      pathname: `https://mainnet-public.mirrornode.hedera.com/api/v1${path}`,
```
the url domain left part need to be an env in env.js

gre commented on src/families/hedera/js-scanAccounts.ts:
```diff
@@ -0,0 +1,126 @@
+}): Observable<ScanAccountEvent> {
+  var derivationMode = scheme ?? getDerivationModesForCurrency(currency)[0];
+
+  return new Observable((o) => {
```
In future, it will be preferrable to use jsHelpers which should help you generalizing the 'scanAccounts' and 'sync'. Is there any reason it was not used yet?

gre commented on src/types/account.ts:
```diff
@@ -133,6 +133,9 @@ export type Account = {
   index: number;
+  // a "virtual" address is a non-verified address
+  // setting this triggers the receive flow to show a warning that it can't verify the address
+  virtualAddress?: string;
```
adding new fields at top level is costly. If there is really a need to have this, we can discuss, but generally speaking we would just introduce a namespaced `hederaResources` object (like for tronResources, bitcoinResources, etc..)

gre reviewed (commented):
hot feedbacks on this first iteration.

mehcode committed:
hedera: remove virtualAddress in favor of hederaResources.accountId

mehcode committed:
hedera: move mirror API URL to env.ts

mehcode committed:
hedera: update @hashgraph/sdk

mehcode committed:
hedera: handle hederaResources in asReusltAccount in cross.ts

mehcode committed:
hedera: hide system accounts in recipients and only show node if this is a node payment

mehcode committed:
hedera: use transaction ID format that works in the explorer

mehcode committed:
update hbar.svg from icon by ledger

mehcode committed:
hedera: run prettier and replace let with const where possible (lint)

mehcode committed:
hedera: remove unused imports, run prettier on one more file (lint)

mehcode committed:
hedera: fix getOperationsForAccount, off by 1

gre reviewed (changes requested):
some feedback on our side.
do you think it's also possible to add some minimalistic tests? we can guide you through it if necessary, but see https://developers.ledger.com/docs/coin/run-ledger-live/

it's necessary that our coins are tested so we can have non regression automated from the beginning.

haammar-ledger commented on src/families/hedera/utils.ts:
```diff
@@ -0,0 +1,29 @@
+  transaction: Transaction;
+}): { amount: BigNumber; estimatedFees: BigNumber; totalSpent: BigNumber } {
+  // NOTE: Hedera declares stable fees in USD
+  //       If we can get the current USD/HBAR price here..
```
Question for @gre: do you think we can get the countervalue here?

haammar-ledger replied:
As discussed during our sync meeting, please keep displaying fees in coin unit, as Ledger Live isn't designed to display fees in fiat unit.

jackthta replied:
👍🏻

haammar-ledger commented on src/families/hedera/js-estimateMaxSpendable.ts:
```diff
@@ -0,0 +1,28 @@
+import type { Account, AccountLike } from "../../types";
+import type { Transaction } from "./types";
+
+export default function estimateMaxSpendable({
```
This code is very similar to the `calculateAmount` function in `utils.ts`, can you maybe use it there?

jackthta replied:
👍🏻

haammar-ledger commented on src/families/hedera/js-receive.ts:
```diff
@@ -0,0 +1,65 @@
+import { Account, isSegwitDerivationMode } from "../../types";
+import getAddress from "../../hw/getAddress";
+
+export default function receive(
```
Same concern as for the scan & sync account, the default receive should be used preferably... but it's on us to help you find a way to fit our model.

haammar-ledger commented on src/families/hedera/js-transaction.ts:
```diff
@@ -0,0 +1,46 @@
+  transaction: Transaction,
+  patch: Partial<Transaction>
+): Transaction {
+  return { ...transaction, ...patch } as unknown as Transaction;
```
I think you don't need this double cast?

jackthta replied:
👍🏻

haammar-ledger commented on src/families/hedera/js-sync.ts:
```diff
@@ -0,0 +1,49 @@
+import { getOperationsForAccount } from "./api/mirror";
+import { getAccountBalance } from "./api/network";
+
+export default function sync(
```
This implementation is not optimal as the full list of operations is replaced every time.
You can consider using `mergeOps`, see examples in other coin families.

There may be other concerns with not using the generic helpers, as already raised by @gre. We need to discuss and help you on that topic.

haammar-ledger replied:
To follow-up on this topic, I _think_ we are ok with having custom `sync` and `scanAccount` logics, to be confirmed by @gre.

But at minimum we need to make sure the performance is acceptable. The optimization issue that I mentioned above should be addressed.

haammar-ledger replied:
Digging a bit more, I'm wondering if there is a way to use the standard mechanism, by using `ChildAccount`... I need to think more about it.

haammar-ledger replied:
Never mind previous comment, probably a dead-end.

jackthta replied:
👍🏼

haammar-ledger commented on src/families/hedera/js-getTransactionStatus.ts:
```diff
@@ -0,0 +1,55 @@
+      );
+
+      if (senderAccountId.equals(recipientAccountId)) {
+        errors.recipient = new InvalidAddressBecauseDestinationIsAlsoSource("");
```
Is it forbidden to self-send with Hedera?

jackthta replied:
To the best of my knowledge, yes it is forbidden to self-send.

gre commented on src/derivation.ts:
```diff
@@ -221,6 +221,9 @@ const modes = Object.freeze({
     overridesDerivation: "44'/461'/0'/0/<account>",
   },
+  hederaBip44: {
+    overridesDerivation: "44/3030/0/0/<account>",
```
44'/3030' ?

jackthta replied:
I'm not sure why it's the case, but `44'/3030'/0/0/<account>` doesn't seem to work or when trying to add an account, will not return the associated list of accounts. `44/3030'/0/0/<account>` _does_ work, but I'm not sure how much of an improvement it would be over what's currently there.

gre commented on src/families/hedera/hw-app-hedera/index.ts:
```diff
@@ -0,0 +1,103 @@
+};
+
+/** Hedera BOLOS API */
+export default class Hedera {
```
i think this can be done in second time once the work is validated into our source code. topic for us later. not blocker.

steven-sheehy commented on src/env.ts:
```diff
@@ -125,6 +125,11 @@ const envDefinitions = {
     desc: "proxy url for trongrid API",
   },
+  API_HEDERA_MIRROR: {
+    def: "https://mainnet-public.mirrornode.hedera.com",
```
This should be updated to use the unthrottled mirror node:

```suggestion
    def: "https://mainnet.mirrornode.hedera.com",
```

jackthta replied:
Changing the mirror api url to `https://mainnet.mirrornode.hedera.com` makes the app stall indefinitely on the synchronizing step (where the url is queried to check for account associated with the ledger device). I changed it back to the previous url for the time being until this is addressed.

jackthta committed:
test: add unit tests

Signed-off-by: Jack Ta <jack.th.ta@gmail.com>

jackthta committed:
refactor: remove double type assertion in js-transaction.ts

Signed-off-by: Jack Ta <jack.th.ta@gmail.com>

jackthta committed:
refactor: use estimateMaxSpendable() in utils.ts

Signed-off-by: Jack Ta <jack.th.ta@gmail.com>

jackthta commented on src/families/hedera/api/mirror.ts:
```diff
@@ -0,0 +1,136 @@
+      date: timestamp,
+      blockHeight: 1,
+      // NOTE: not sure what to put here, there are no "blocks" in hedera
+      blockHash: raw.transaction_hash,
```
👍🏻

jackthta commented on src/families/hedera/api/network.ts:
```diff
@@ -0,0 +1,54 @@
+let _hederaClient: hedera.Client | null = null;
+
+function getClient(): hedera.Client {
+  _hederaClient ??= hedera.Client.forMainnet().setMaxNodesPerTransaction(1);
```
The client may be targeting a range from node `0.0.3` to `0.0.20` and I don't believe there is a way to provide a custom URL.

I may need an expert opinion to confirm my assumptions though.

haammar-ledger replied:
Topic still pending, we should discuss it during our sync meeting. Either a reverse proxy isn't considered mandatory on this part, or you could maybe target the network directly, same way as it's done in `mirror.ts`.

jackthta replied:
Apparently we _can_ set a proxy via `client.setNetwork()`, but I wanted to ask whether the proxy has been set up? If so, can it be provided so I can try it out to see if it can work please

haammar-ledger replied:
Cool, yes we now have a proxy up and running : `https://hedera.coin.ledger.com`

jackthta commented on src/families/hedera/api/mirror.ts:
```diff
@@ -0,0 +1,136 @@
+export async function getAccountsForPublicKey(
+  publicKey: string
+): Promise<Account[]> {
+  const r = await fetch("/accounts", { "account.publicKey": publicKey });
```
I just want to double check if I'm understanding this correctly, but would it be something like this?

`const r = await fetch("/accounts", { "account.publicKey": publicKey, "balance": false });`

steven-sheehy replied:
Correct. We have [swagger docs](https://mainnet.mirrornode.hedera.com/api/v1/docs/#/) you can also check. Setting this param will speed up the response time.

Could also put a `limit` param on it if you know you don't care about multiple accounts being associated with the same key. That would also speed it up.

jackthta replied:
Thank you! I will make this edit 👍🏻

haammar-ledger replied:
It seems from our tests that the `account.balance=false` path param currently doesn't work?

```
$ curl "https://mainnet-public.mirrornode.hedera.com/api/v1/accounts?account.publicKey=9e92a312233d5fd6b5a723875aeea2cea81a8e48ffc00341cff6dffcfd3ab7f2&account.balance=false"
{"_status":{"messages":[{"message":"Invalid parameter: account.balance"}]}}
```

```
$ curl "https://mainnet-public.mirrornode.hedera.com/api/v1/accounts?account.publicKey=9e92a312233d5fd6b5a723875aeea2cea81a8e48ffc00341cff6dffcfd3ab7f2"
{"accounts":[],"links":{"next":null}}
```

steven-sheehy replied:
It's just `balance`, not `account.balance`. Again, please reference the [swagger](https://mainnet.mirrornode.hedera.com/api/v1/docs/#/) site.

jackthta replied:
That was my mistake, I looked at the swagger doc and assumed by context that it would be `account.balance`. It should be fixed now.

jackthta committed:
test: fix utils.test.ts

Signed-off-by: Jack Ta <jack.th.ta@gmail.com>

jackthta committed:
refactor: set block properties to null

Signed-off-by: Jack Ta <jack.th.ta@gmail.com>

jackthta committed:
refactor: optimize getAccountsForPublicKey() by setting balance param to false

Signed-off-by: Jack Ta <jack.th.ta@gmail.com>

haammar-ledger commented on src/cross.ts:
```diff
@@ -20,6 +25,8 @@ export type AccountData = {
   index: number;
   balance: string;
+  // chain-specific resources
+  hederaResources?: HederaResourcesRaw;
```
@jackthta According to my understanding, the `accountId` could be retrieved on live-mobile side, in the same way it is retrieved on live-desktop, from the public key, which should be already provided in the AccountData. If you can please confirm, this would be a way to remove the need for `hederaResources` here.

jackthta commented on src/env.ts:
```diff
@@ -125,6 +125,11 @@ const envDefinitions = {
     desc: "proxy url for trongrid API",
   },
+  API_HEDERA_MIRROR: {
+    def: "https://mainnet-public.mirrornode.hedera.com",
```
Using the ledger reverse-proxy here works 👍🏻. I'll push up the edit once I figure out how to set the reverse-proxy in `network.ts` (running into some difficulty using the aforementioned `client.setNetwork()` function).

jackthta replied:
👋🏻 @haammar-ledger

I've tried setting the `hedera.Client` to target `https://hedera.coin.ledger.com`, but it doesn't seem to work. I'm not sure why, but perhaps because `hedera.Client.forMainnet()` sets into the client a network of nodes (which are the mainnet nodes). 

This is the structure/how it looks:
> MAINNET: {
        "35.237.200.180:50211": new AccountId(3),
        "35.186.191.247:50211": new AccountId(4),
        "35.192.2.25:50211": new AccountId(5),
        "35.199.161.108:50211": new AccountId(6),
        "35.203.82.240:50211": new AccountId(7),
        "35.236.5.219:50211": new AccountId(8),
        "35.197.192.225:50211": new AccountId(9),
        "35.242.233.154:50211": new AccountId(10),
        "35.240.118.96:50211": new AccountId(11),
        "35.204.86.32:50211": new AccountId(12),
        "35.234.132.107:50211": new AccountId(13),
        "35.236.2.27:50211": new AccountId(14),
        "35.228.11.53:50211": new AccountId(15),
        "34.91.181.183:50211": new AccountId(16),
        "34.86.212.247:50211": new AccountId(17),
        "172.105.247.67:50211": new AccountId(18),
        "34.89.87.138:50211": new AccountId(19),
        "34.82.78.255:50211": new AccountId(20),
    }

👆🏻 [Source](https://github.com/hashgraph/hedera-sdk-js/blob/f48dcb2bb2a596fae5fca7eea6c76a862a446ee4/src/client/NodeClient.js#L37-L56)

I'm having a difficult time seeing how to target the network directly.. but if you have any suggestions, please do let me know! Otherwise, I hope it will be okay to leave it as is.

jackthta committed:
chore: update @hashgraph/sdk

Signed-off-by: Jack Ta <jack.th.ta@gmail.com>

jackthta committed:
fix: fix invalid recipient address error for Send flow

Signed-off-by: Jack Ta <jack.th.ta@gmail.com>

jackthta committed:
fix: fix looping prompt bug during sign transaction step for Send flow

Reason(s) for changing `freezeWith(getClient())` with `freeze()` & `.setNodeAccountIds([new AccountId(3)])`:
- Using `freezeWith(getClient())` where `getClient()` is `hedera.Client.forMainnet()` OR `hedera.Client.forNetwork({MAINNET_NETWORK})` will create a bug where the `TransferTransaction` will have 18 members in both its `_nodeIds` and `_signedTransactions`.
  - The 18 members in `_nodeIds` are nodes `0.0.3` ~> `0.0.20` and the 18 members in `_signedTransactions` are one of the same of the body bytes to be signed.
  - The problem with this bug is that it will loop prompt the user 18 times on the ledger device to confirm the transaction and once that's exhaustedly done, it inevitably fails.
- Setting `setNodeAccountIds()` to an array containing more than one `AccountId` will yield the same prompt loop problem mentioned above.

jackthta committed:
feat: show verification info on sign transaction step

Signed-off-by: Jack Ta <jack.th.ta@gmail.com>

jackthta committed:
refactor: optimize sync operation

Signed-off-by: Jack Ta <jack.th.ta@gmail.com>

haammar-ledger commented on src/families/hedera/utils.ts:
```diff
@@ -0,0 +1,34 @@
+
+  const amount =
+    transaction.useAllAmount == true
+      ? // as fees are based on a currency conversion, we stay
+        // on the safe side here and double the estimate for "max spendable"
```
obsolete comment, please remove

jackthta committed:
docs: remove obsolete comment

Signed-off-by: Jack Ta <jack.th.ta@gmail.com>

jackthta committed:
fix: live mobile add acc flow's grpc error

Signed-off-by: Jack Ta <jack.th.ta@gmail.com>

jackthta committed:
wip: add hedera bot

Signed-off-by: Jack Ta <jack.th.ta@gmail.com>

jackthta committed:
wip(hedera bot): finish `speculos-deviceActions` file

Signed-off-by: Jack Ta <jack.th.ta@gmail.com>

jackthta committed:
fix(hedera bot): correct `speculos-deviceActions`

Signed-off-by: Jack Ta <jack.th.ta@gmail.com>

jackthta committed:
refactor: `speculos-deviceActions` to remove `expectedValue` field in Fee step

Signed-off-by: Jack Ta <jack.th.ta@gmail.com>

jackthta committed:
fix: add missing 'Max Fee' step to `speculos-deviceActions`

Signed-off-by: Jack Ta <jack.th.ta@gmail.com>

jackthta committed:
refactor: update specs.ts

Signed-off-by: Jack Ta <jack.th.ta@gmail.com>

jackthta committed:
fix(hedera bot): add missing amount for send max mutation

Signed-off-by: Jack Ta <jack.th.ta@gmail.com>

jackthta committed:
refactor: use `calculateAmount` in `prepareTransaction` instead of in bot test logic

Signed-off-by: Jack Ta <jack.th.ta@gmail.com>

jackthta committed:
fix(hedera bot): set expected value of Max Fee step to 1 hbar

Signed-off-by: Jack Ta <jack.th.ta@gmail.com>

jackthta committed:
test: fix unit tests

Signed-off-by: Jack Ta <jack.th.ta@gmail.com>

haammar-ledger commented:
Replaced with https://github.com/LedgerHQ/ledger-live-common/pull/1887
